File: sbc/src/logging/sinks/file_sink.cpp
#include "file_sink.h"
#include <filesystem>
#include <iostream>

namespace fs = std::filesystem;

FileSink::FileSink(const std::string& file_path, size_t max_size_mb, int max_rotated)
    : file_path_(file_path), max_size_bytes_(max_size_mb * 1024 * 1024), max_rotated_(max_rotated) {
    openFile();
}

FileSink::~FileSink() {
    closeFile();
}

void FileSink::write(const json& log_entry) {
    std::lock_guard<std::mutex> lock(mutex_);

    if (!file_.is_open()) {
        openFile();
    }

    // Write as JSON Lines format (one JSON object per line)
    file_ << log_entry.dump() << '\n';

    // Check if rotation is needed
    if (getFileSize() >= max_size_bytes_) {
        rotateFile();
    }
}

void FileSink::flush() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (file_.is_open()) {
        file_.flush();
    }
}

std::string FileSink::name() const {
    return "file:" + file_path_;
}

void FileSink::openFile() {
    // Create parent directory if it doesn't exist
    fs::path file_path(file_path_);
    if (file_path.has_parent_path()) {
        fs::create_directories(file_path.parent_path());
    }

    file_.open(file_path_, std::ios::out | std::ios::app);
    if (!file_.is_open()) {
        std::cerr << "FileSink: Failed to open log file: " << file_path_ << std::endl;
    }
}

void FileSink::closeFile() {
    if (file_.is_open()) {
        file_.flush();
        file_.close();
    }
}

void FileSink::rotateFile() {
    closeFile();

    // Rotate existing files: file.2 -> file.3, file.1 -> file.2, file -> file.1
    for (int i = max_rotated_ - 1; i >= 1; i--) {
        std::string old_path = file_path_ + "." + std::to_string(i);
        std::string new_path = file_path_ + "." + std::to_string(i + 1);

        if (fs::exists(old_path)) {
            if (i == max_rotated_ - 1) {
                // Delete oldest file
                fs::remove(old_path);
            } else {
                // Rename file
                fs::rename(old_path, new_path);
            }
        }
    }

    // Rename current file to .1
    if (fs::exists(file_path_)) {
        std::string backup_path = file_path_ + ".1";
        fs::rename(file_path_, backup_path);
    }

    // Open new file
    openFile();
}

size_t FileSink::getFileSize() {
    if (!file_.is_open()) {
        return 0;
    }

    // Get current position
    auto pos = file_.tellp();
    if (pos < 0) {
        // If tellp() fails, check file size on disk
        if (fs::exists(file_path_)) {
            return fs::file_size(file_path_);
        }
        return 0;
    }

    return static_cast<size_t>(pos);
}

File: android/app/src/main/java/uk/unmannedsystems/dpm_android/logging/sinks/FileSink.kt
package uk.unmannedsystems.dpm_android.logging.sinks

import uk.unmannedsystems.dpm_android.logging.LogEntry
import uk.unmannedsystems.dpm_android.logging.StructuredLogger
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException

/**
 * File Sink - Writes logs to JSON Lines files with size-based rotation
 *
 * Features:
 * - JSON Lines format (.jsonl)
 * - Size-based rotation (default 50 MB)
 * - Keeps N rotated files (default 3)
 * - Thread-safe writes
 *
 * File naming:
 * - Current: ground-side.jsonl
 * - Rotated: ground-side.1.jsonl, ground-side.2.jsonl, ground-side.3.jsonl
 */
class FileSink(
    private val logDir: File,
    private val maxFileSizeMB: Int = 50,
    private val maxRotatedFiles: Int = 3
) : LogSink {
    private val TAG = "FileSink"

    private val baseFileName = "ground-side.jsonl"
    private val maxFileSizeBytes = maxFileSizeMB * 1024 * 1024L

    private var currentFile: File = File(logDir, baseFileName)
    private var writer: BufferedWriter? = null

    init {
        // Ensure log directory exists
        if (!logDir.exists()) {
            logDir.mkdirs()
            StructuredLogger.info("STORAGE", "Created log directory: ${logDir.absolutePath}")
        }

        // Open the current log file
        try {
            writer = BufferedWriter(FileWriter(currentFile, true))
            StructuredLogger.info("STORAGE", "FileSink initialized: ${currentFile.absolutePath}")
        } catch (e: IOException) {
            StructuredLogger.error("STORAGE", "Failed to open log file: ${e.message}")
        }
    }

    /**
     * Write a log entry to the file
     */
    @Synchronized
    override fun write(entry: LogEntry) {
        try {
            // Check if rotation is needed
            if (currentFile.length() >= maxFileSizeBytes) {
                rotate()
            }

            // Write JSON line
            writer?.write(entry.toJson())
            writer?.newLine()
        } catch (e: IOException) {
            StructuredLogger.error("STORAGE", "Failed to write log entry: ${e.message}")
        }
    }

    /**
     * Flush buffered writes to disk
     */
    @Synchronized
    override fun flush() {
        try {
            writer?.flush()
        } catch (e: IOException) {
            StructuredLogger.error("STORAGE", "Failed to flush log file: ${e.message}")
        }
    }

    /**
     * Close the file sink
     */
    @Synchronized
    override fun close() {
        try {
            writer?.flush()
            writer?.close()
            StructuredLogger.info("STORAGE", "FileSink closed")
        } catch (e: IOException) {
            StructuredLogger.error("STORAGE", "Failed to close log file: ${e.message}")
        }
    }

    /**
     * Rotate log files
     *
     * Process:
     * 1. Close current writer
     * 2. Delete oldest rotated file (if exists)
     * 3. Rename all rotated files (increment number)
     * 4. Rename current file to .1.jsonl
     * 5. Create new current file
     * 6. Open new writer
     */
    private fun rotate() {
        StructuredLogger.info("STORAGE", "Rotating log files (current size: ${currentFile.length() / 1024 / 1024} MB)")

        try {
            // Close current writer
            writer?.flush()
            writer?.close()

            // Delete oldest file (e.g., ground-side.3.jsonl)
            val oldestFile = File(logDir, "$baseFileName.$maxRotatedFiles.jsonl")
            if (oldestFile.exists()) {
                oldestFile.delete()
                StructuredLogger.debug("STORAGE", "Deleted oldest log file: ${oldestFile.name}")
            }

            // Rename existing rotated files (increment number)
            for (i in (maxRotatedFiles - 1) downTo 1) {
                val sourceFile = File(logDir, "$baseFileName.$i.jsonl")
                val destFile = File(logDir, "$baseFileName.${i + 1}.jsonl")
                if (sourceFile.exists()) {
                    sourceFile.renameTo(destFile)
                    StructuredLogger.debug("STORAGE", "Renamed ${sourceFile.name} -> ${destFile.name}")
                }
            }

            // Rename current file to .1.jsonl
            val rotatedFile = File(logDir, "$baseFileName.1.jsonl")
            currentFile.renameTo(rotatedFile)
            StructuredLogger.debug("STORAGE", "Renamed ${currentFile.name} -> ${rotatedFile.name}")

            // Create new current file
            currentFile = File(logDir, baseFileName)
            writer = BufferedWriter(FileWriter(currentFile, true))
            StructuredLogger.info("STORAGE", "Created new log file: ${currentFile.name}")
        } catch (e: IOException) {
            StructuredLogger.error("STORAGE", "Failed to rotate log files: ${e.message}")
        }
    }
}
